pubtools/_quay/quay_client.py

```diff
@@ -0,0 +1,192 @@
+import json
+import logging
+import requests
+from requests.packages.urllib3.util.retry import Retry
+from six.moves.urllib.request import parse_http_list, parse_keqv_list
+
+from .exceptions import ManifestTypeError, RegistryAuthError
+from .quay_session import QuaySession
+
+LOG = logging.getLogger()
+LOG.setLevel(logging.INFO)
+
+
+class QuayClient:
+    """Class for performing Docker HTTP API operations with the Quay registry."""
+
+    MANIFEST_LIST_TYPE = "application/vnd.docker.distribution.manifest.list.v2+json"
+
+    def __init__(self, username, password, host=None):
+        """
+        Initialize.
+
+        Args:
+            username (str):
+                Quay username.
+            password (str):
+                Quay password.
+            host (str):
+                Quay registry URL.
+        """
+        self.username = username
+        self.password = password
+        self.session = QuaySession(hostname=host)
+
+    def get_manifest_list(self, image, raw=False):
+        """
+        Get manifest list of a given image. Throw an error if it couldn't be retrieved.
+
+        Args:
+            image (str):
+                Image for which to get the manifest list.
+            raw (bool):
+                Whether to return the manifest as raw JSON.
+        Returns (dict|str):
+            The returned manifest list.
+        Raises:
+            ManifestTypeError:
+                When the image doesn't have a manifest list.
+        """
+        LOG.info("Getting manifest list of image {0}".format(image))
+        repo, ref = self._parse_and_validate_image_url(image)
+        endpoint = "{0}/manifests/{1}".format(repo, ref)
+        # request 'Content-Type' to be manifest list
+        kwargs = {"headers": {"Accept": QuayClient.MANIFEST_LIST_TYPE}}
+
+        response = self._request_quay("GET", endpoint, kwargs)
+        if response.headers["Content-Type"] != QuayClient.MANIFEST_LIST_TYPE:
+            raise ManifestTypeError(
+                "Image {0} doesn't have a manifest list".format(image)
+            )
+
+        if raw:
+            return response.text
+        else:
+            return response.json()
+
+    def upload_manifest_list(self, manifest_list, image):
+        """
+        Upload a manifest list to a specified image.
+
+        Args:
+            manifest_list (dict):
+                Manifest list to be uploaded.
+            image (str):
+                Image address to upload the manifest list to.
+        """
+        LOG.info("Uploading manifest list to image {0}".format(image))
+        repo, ref = self._parse_and_validate_image_url(image)
+        endpoint = "{0}/manifests/{1}".format(repo, ref)
+        kwargs = {
+            "headers": {"Content-Type": QuayClient.MANIFEST_LIST_TYPE},
+            "data": json.dumps(manifest_list),
```

Comment:
Could this have 'indent' added so manifest list can be pretty-printed?

Reply:
Yep, that's a good idea